### PR TITLE
Stroop Updates - replace RandomizeColoredTextOrBox with Random Text Underlining

### PR DESCRIPTION
**Pull request by eschramm**
This PR looks to accomplish 2 goals:

1. After review with the client's psychologists, the intent here is to force "context switching". Switching between the box and the text does not do this. Instead, the recommendation is to randomly (50 %) underline the colored text and instruct the user to select the word displayed instead of the color of the word.

2. Allow for a specific test to be shown to the user for a practice test before the actual test - e.g., show one case of congruent color/text and then one that is incongruent, and finally one that is incongruent and underlined. This also will allow testing of functionality and results (see https://github.com/CareEvolution/CEVResearchKit/pull/204)

This should require no changes on the server to handle the updated StroopResult.

Tests added to ensure randomization works as expected.

### CODE REVIEW

File: ResearchKit/ActiveTasks/ORKStroopStepViewController.h [hunk 1]
```diff
@@ -39,6 +39,9 @@ NS_ASSUME_NONNULL_BEGIN
 ORK_CLASS_AVAILABLE
 @interface ORKStroopStepViewController : ORKActiveStepViewController
 
+// public for testing
+- (BOOL)randomBoolWithTrueProbability:(NSNumber * __nonnull)trueProbability;
```
Comment (overcyn) [inline ResearchKit/ActiveTasks/ORKStroopStepViewController.h]: Would this be a better fit in ORKHelpers_Internal.h?

File: ResearchKit/ActiveTasks/ORKStroopStep.m [hunk 2]
```diff
@@ -73,27 +117,43 @@ - (BOOL)startsFinished {
...
+    step.useGridLayoutForButtons = self.useGridLayoutForButtons;
+    step.nonRandomizedTests = self.nonRandomizedTests;
     return step;
 }
 
 - (instancetype)initWithCoder:(NSCoder *)aDecoder {
     self = [super initWithCoder:aDecoder];
     if (self ) {
         ORK_DECODE_INTEGER(aDecoder, numberOfAttempts);
+        ORK_DECODE_OBJ(aDecoder, probabilityOfVisualAndColorAlignment);
```
Comment (overcyn) [inline ResearchKit/ActiveTasks/ORKStroopStep.m:203]: Does ORKStroopColor and ORKStroopTest need to implement NSCoder?

File: ResearchKit/ActiveTasks/ORKStroopStep.m [hunk 3]
```diff
@@ -73,27 +117,43 @@ - (BOOL)startsFinished {
...
 - (BOOL)isEqual:(id)object {
     BOOL isParentSame = [super isEqual:object];
     
     __typeof(self) castObject = object;
-    return (isParentSame && (self.numberOfAttempts == castObject.numberOfAttempts));
+    return (isParentSame && (self.numberOfAttempts == castObject.numberOfAttempts)
+            && [self.probabilityOfVisualAndColorAlignment isEqual:castObject.probabilityOfVisualAndColorAlignment]
+            && (self.stroopStyle == castObject.stroopStyle)
+            && (self.useGridLayoutForButtons == castObject.useGridLayoutForButtons)
+            && [self.nonRandomizedTests isEqual:castObject.nonRandomizedTests]);
```
Comment (overcyn) [inline ResearchKit/ActiveTasks/ORKStroopStep.m]: Also does ORKStroopColor and ORKStroopTest need to implement `isEqual:`?

File: ResearchKit/ActiveTasks/ORKStroopStep.h [hunk 4]
```diff
@@ -31,34 +31,82 @@
...
+ By default, this property is set to `NO`
 */
 @property (nonatomic, assign) BOOL useGridLayoutForButtons;
 
+/**
+ When supplied, will run the test using the characteristics of this array. This aids in testing or
+ to provide a specific practice example set. If set, this overrides the randomizeVisualAndColorAlignment,
+ stroopStyle and numberOfAttempts properties.
+ */
+@property (nonatomic, strong, nullable) NSArray<ORKStroopTest *> *nonRandomizedTests;
```
Comment (overcyn) [inline ResearchKit/ActiveTasks/ORKStroopStep.h]: NonRandomizedTests reads oddly to me (ie `stroopStep.nonRandomizedTests = [test1, test2, ...]`). What about just `stroopTests`?

**Review by overcyn — APPROVED** [review]
Looks really good. Minor comments inline.

File: ResearchKit/ActiveTasks/ORKStroopStepViewController.h (hunk 1 above)
Comment (eschramm) [reply ResearchKit/ActiveTasks/ORKStroopStepViewController.h]: Thanks for the suggestion. I really like the updated approach now which theoretically could be reused in other places in ResearchKit for randomization.

File: ResearchKit/ActiveTasks/ORKStroopStep.m (hunk 2 above)
Comment (eschramm) [reply ResearchKit/ActiveTasks/ORKStroopStep.m:203]: Hmm, probably. I see that we should also be adhering to `NSCopying` as well and have added that, too.

File: ResearchKit/ActiveTasks/ORKStroopStep.h (hunk 4 above)
Comment (eschramm) [reply ResearchKit/ActiveTasks/ORKStroopStep.h]: Agreed, this sounds much cleaner as `stroopTests`.

**Comment by overcyn** [comment]
LGTM

**Pull request merged**